### DISCUSSION TITLE
How to use and create hash key for update server

### DISCUSSION THREAD

**Discussion by jschmi102** [joomla/joomla-cms#37114] [General]
With joomla 4 you may use a hash key for updates. You will have to define it in you update.xml file. I tried this with using a key created by php's "hash_file" function but the created key caused a failure in the update with "invalid verification of key".
So far I did not find which tool to use to create correct keys. which can be used with a joomla update server - can anybody advise?

**Comment by alikon** [comment]
one of this will work sha256,sha384,sha512

**Comment by counterpoint** [comment]
If you are running on Linux and have access to the command line, you can find the SHA256 hash by:
      sha256sum mypackage.zip
and see an answer like:
      7f3d59eacebe92fbce60eb1f8b0b5795c19bbbce3c2975e9fcdab3a0f78a8720 mypackage.zip
Copy just the hash into the <sha256> section of the XML.

**Comment by jschmi102** [comment]
hi,

I've created a web-app to create hashes for my extension updates. It uses php-function 
 ` hash_file()) ` e.g. `hash_file("sha256", $zipfile, false)`

If you are interested I'll send my solution to you.

**Reply by alikon** [reply]
yes please share

**Reply by dgrammatiko** [reply]
You definitely don't need PHP for this...

**Comment by dgrammatiko** [comment]
FWIW any of the modern browsers could help with the hashes and here's an implementation: https://hash-it.dgrammatiko.dev

The source code is here: https://github.com/dgrammatiko/hash-it

**Reply by jschmi102** [reply]
hi,
I prefer doing it programmatically - rather than doing manually:

1. find/call online webservice
2.  upload zip-file
3.  create hash
4.  copy result
5.  edit update server xml
6.  paste result

If your extension has 4 zip-files (e.g. for package, component, plugin and module) you run into a time consuming process.

**Reply by wojsmol** [reply]
@jschmi102 You are using UNIX like system or Windows?

**Reply by jschmi102** [reply]
hi,
my development system is mainly on windows 11. Target platform is using  linux  - the provided app runs with joomla on both systems. The only difference is path-settings in file settings.ini.

**Comment by jschmi102** [comment]
hi,
here is my code - Unpack files into a sub-directory of your joomla installtion (e.g. "php-apps"). Intstructions to install is provided in file readme.txt.   

[update_hashes.zip](https://github.com/joomla/joomla-cms/files/9505977/update_hashes.zip)

**Comment by alikon** [comment]
thank you both
